Add Texpacker

silverweed opened the pull request:
(See #12)
Texpacker now works hopefully well. I tried sticking to the rest of the tools' style as much as possible, so it should be pretty familiar to review.

### Usage
```
$ ./texpacker -h
Usage: ./texpacker [flags] <file1> [<file2> ...]
  -noheader
    	Don't write header, output raw PNG
  -o string
    	Output file (- for STDOUT) (default "-")
  -strip string
    	Prefix to strip from file paths for hashing (default <cwd>)

Supported input formats:
    JPEG
    GIF
    PNG
```
Basically it takes input textures as positional arguments and spits out a binary file with the format `[header | PNG image]`. It can be told to output just the PNG for debugging purposes.
The `-strip` flag is used to make the tool usable from any directory: the hashed paths will be relative to the path given to `-strip`.

The packing algorithm is currently maxrects and uses  `"github.com/adinfinit/texpack/maxrect"` which is MIT-licensed.

### Header format
The header looks like this:
```
ParentTexture Hashed Path [4B]
Entry Count               [4B]
Entry0                    [12B]
...

where Entry is:
Hashed Path [4B]
StartX      [2B]
StartY      [2B]
SizeX       [2B]
SizeY       [2B]
```

---

Forgot anything?

silverweed committed:
implement basic texture packing

silverweed committed:
make it work with maxrect

silverweed committed:
add texpacker to build.{sh,cmd}

silverweed committed:
add header outputting

silverweed committed:
fix PutUint's in format.go

silverweed committed:
add noheader flag

silverweed committed:
add prefix stripping option to texpacker

silverweed committed:
small fix to usage

silverweed committed:
forgot to relativize outpath too

Hamcha commented on tools/build.sh:
```diff
@@ -2,4 +2,5 @@ go build -o bin/objconv ./objconv
-go build -o bin/tevasm ./tevasm
\ No newline at end of file
+go build -o bin/tevasm ./tevasm
+go build -o bin/texpacker ./texpacker
```
Please add the "go get" command on top of both build files, just like tools-extra/build.*

zeroZshadow commented on tools/texpacker/format.go:
```diff
@@ -0,0 +1,58 @@
+
+// 8 bytes
+type Rect struct {
+	Start, Size Point
```
Size isn't a point, so either rename `Point` to be less specialized (Vector2?)
Or change what Size is?

Hamcha reviewed (changes requested):
This is still using a single file for output while it should do two (packed PNG + metadata). My original doc referred to "Header" meaning "C header file", not header of the actual texture.

zeroZshadow commented on tools/texpacker/main.go:
```diff
@@ -0,0 +1,66 @@
+	noheader := flag.Bool("noheader", false, "Don't write header, output raw PNG")
+	cwd, err := os.Getwd()
+	checkErr(err, "Couldn't get working directory")
+	stripPfx := flag.String("strip", cwd, "Prefix to strip from file paths for hashing")
```
Why not call this `-prefix`? The current name sounds like its a toggle flag

zeroZshadow commented on tools/texpacker/main.go:
```diff
@@ -0,0 +1,66 @@
+	// All positional arguments are input files
+	inputFiles := flag.Args()
+
+	maxBounds := image.Point{1 << 16, 1 << 16} // TODO
```
Forgot a TODO here, flag?

zeroZshadow commented on tools/texpacker/format.go:
```diff
@@ -0,0 +1,58 @@
+func ToFileHash(s string) FileHash {
+	hash.Reset()
+	_, err := fmt.Fprintf(hash, s)
+	if err != nil {
```
Why doesnt this use checkErr ?

zeroZshadow commented on tools/texpacker/main.go:
```diff
@@ -0,0 +1,66 @@
+		checkErr(err, "Error converting file path %s to absolute", path)
+		packer.Add(abspath)
+	}
+	if err := packer.Save(); err != nil {
```
Why doesn this use checkErr ?

zeroZshadow commented on tools/texpacker/packer.go:
```diff
@@ -0,0 +1,245 @@
+
+// writeImageAt adds all pixels of `src` to `dst`, starting from pixel at (`x`,`y`).
+// It returns an error if the whole source image couldn't be added to the target.
+func writeImageAt(dst *image.RGBA, src image.Image, x, y int) error {
```
This could instead use `image.SubImage` and the `image/draw` functions?

zeroZshadow commented on tools/texpacker/packer.go:
```diff
@@ -0,0 +1,245 @@
+			if tryOk {
+				contextSize = trySize
+				rects = tryRects
+				shrunk = true
```
This will still cause it to try shrinkY, is that ok?

silverweed replied:
That's probably a typo in the function I copy-pasted, it should be `shrinkY = false` in the `shrinkY` branch.

silverweed committed:
fix some issues:

- add go get to build scripts
- change Point -> Vector2
- use checkErr where appropriate
- change -strip -> -prefix
- pass max bounds via flags
- use image/draw functions instead of writeImageAt
- fix typo in minimizeFit

silverweed committed:
output 2 separate files for image and header

silverweed commented:
Ok, shouls have fixed all issues. Now the tool outputs 2 files: `<outfile>` and `<outfile>.atlas`, the latter containing the metadata in the exact same format as described in OP.
I've removed the `-noheader` flag and the possibility to output to stdout, as these would yield ill-defined image/metadata pairs.

zeroZshadow commented on tools/texpacker/main.go:
```diff
@@ -22,11 +22,12 @@ func main() {
-	outpath := flag.String("o", "-", "Output file (- for STDOUT)")
-	noheader := flag.Bool("noheader", false, "Don't write header, output raw PNG")
+	outpath := flag.String("o", "out.png", "Output file")
+	maxW := flag.Int("maxwidth", 1<<16, "Max width of output texture in pixels")
```
Up for discussion, do we wanna support non-square atlasses? if not, only needs 1 flag ;)

Hamcha replied:
No reason to, I think.

zeroZshadow commented on tools/texpacker/packer.go:
```diff
@@ -63,36 +63,32 @@ func (packer *TexPacker) Add(texpath string) error {
 	}
-	return out, nil
+	headerOut, err = os.Create(packer.outfile + ".atlas")
+	return
```
If this returns an error, `TexPacker.Save` will not close the `imageOut`

silverweed committed:
use single flag for size, close imageOut on err

silverweed commented:
Latest commit should have fixed both issues.
